### PR TITLE
Enable configurable API Server Load Balancer

### LINKED ISSUE #486: Private clusters
/kind feature

**Describe the solution you'd like**
[A clear and concise description of what you want to happen.]

Support deploying clusters with no public IP or external load balancer.  For reference: https://github.com/Azure/aks-engine/blob/master/docs/topics/features.md#feat-private-cluster.

Once Azure Bastion is implemented, we should document using Azure Bastion + Private clusters. related to #165

**Anything else you would like to add:**
[Miscellaneous information that will assist in solving the issue.]

**Environment:**

- cluster-api-provider-azure version: 
- Kubernetes version: (use `kubectl version`): 
- OS (e.g. from `/etc/os-release`):

### PR DESCRIPTION
**Pull request by CecileRobertMichon**
**What type of PR is this?**
/kind feature
/kind api-change

**What this PR does / why we need it**: This PR allows configuring the API Server load balancer to enable things like private clusters, eg:

```
  networkSpec:
    apiServerLB:
      type: Internal
```

It adds defaulting + validating webhooks for the new Load Balancer spec. It also allows BYO api server public IP via specifying a pre-existing IP.

Check out the docs to see what this exposes to the user: https://deploy-preview-974--kubernetes-sigs-cluster-api-provider-azure.netlify.app/topics/api-server-endpoint.html

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Fixes #486, #696, #952, #892  

**Special notes for your reviewer**:

Still work in progress, opening to get early feedback.

~Changes are currently breaking.~

_Please confirm that if this PR changes any image versions, then that's the sole change this PR makes._

**TODOs**:

- [x] squash commits
- [x] documentation
- [x] unit tests
- [x] back compat for public API server
- [x] private cluster needs management cluster in the same vnet
- [x] BYO API Server IP delete
- [x] make sure that private control plane VMs have outbound access

**Release note**:

```release-note
Allow configuration of the API Server Load Balancer, including support for private API Server endpoint.
Allow BYO API Server IP
```

### CODE REVIEW

File: api/v1alpha3/types.go
Comment (CecileRobertMichon) [inline api/v1alpha3/types.go:177]: @alexeldeib @MirzaSikander this supersedes #652, along with https://github.com/kubernetes-sigs/cluster-api/issues/3715

**Comment by CecileRobertMichon** [comment]
I still need to fix unit tests and find a way to test private clusters but otherwise, most functionality should be there so this is ready for initial review.

**Comment by CecileRobertMichon** [comment]
/hold for private cluster test

@devigned @alexeldeib @nader-ziada this is ready for initial review

File: templates/flavors/private/patches/private-lb.yaml
Comment (devigned) [inline templates/flavors/private/patches/private-lb.yaml:23]: newline

**Review by devigned — APPROVED** [review]
I really like how this reads: `.spec.controlPlaneEndpoint.host`. Nice work.

Outside of the little nit on newline, lgtm

**Comment by CecileRobertMichon** [comment]
Private cluster creation fails because of the infamous hairpin routing problem with Azure internal LBs, documented [here](https://docs.microsoft.com/en-us/azure/load-balancer/load-balancer-troubleshoot#cause-4-accessing-the-internal-load-balancer-frontend-from-the-participating-load-balancer-backend-pool-vm). 

Essentially, `kubeadm init` fails because it tries to check API Server health using the internal LB private IP from the first control plane VM. Since the VM is the only one in the LB backend at that point, it routes to itself, which fails. This is exactly the same issue described in https://github.com/kubernetes/kubeadm/issues/1685.

I verified that init succeeds when manually changing the controlPlaneEndpoint value to the first control plane VM's private IP.

Potential workarounds are 1) the very hacky iptables duct tape, 2) Azure Application Gateway (?) 3)  HAProxy 3 layers infra setup 4) modify kubeadm to allow using the local endpoint for API server health check during init.

**Comment by vincepri** [comment]
@CecileRobertMichon Where can I get more information about the first two solutions? I'd avoid adding another layer of load balancing if possible. The 4th option might be ok-ish, but we'll need to check with upstream and potentially backport (which mean compatibility matrix for private clusters is going to be different)

**Comment by neolit123** [comment]
> This is exactly the same issue described in kubernetes/kubeadm#1685.

well at least one part of it, that issue ended up as dumping ground for a number of LB problems some of which are not the concern of the kubeadm maintainers.

> modify kubeadm to allow using the local endpoint for API server health check during init.

i'm not objecting to that. response on slack:

> in any case, if we want to change this behavior - let's see it as a change in the KEP for kubeadm HA support first:
https://github.com/kubernetes/enhancements/issues/357
tracking issue in k/kubeadm:
https://github.com/kubernetes/kubeadm/issues/2081

https://kubernetes.slack.com/archives/C2P1JHS2E/p1603291912153700?thread_ts=1603230818.143600&cid=C2P1JHS2E

**Comment by CecileRobertMichon** [comment]
@vincepri the first is to simply run a preKubeadmCommand that sets up an ip table rule `iptables -t nat -A OUTPUT -p all -d "10.128.0.100" -j DNAT --to-destination 127.0.0.1` where `10.128.0.100` is the internal LB private IP.
For the second, see https://docs.microsoft.com/en-us/azure/application-gateway/overview. I don't have a working solution of that one, it's just something that was mentioned as a workaround in Azure docs. I do think it would be more involved and require extra infrastructure (infra costs ++).

Right now what I'm thinking of doing is using the iptables hack to unblock this PR and then work on an upstream proposal to achieve 4) as a long term solution.

**Comment by vincepri** [comment]
> For the second, see https://docs.microsoft.com/en-us/azure/application-gateway/overview. I don't have a working solution of that one, it's just something that was mentioned as a workaround in Azure docs. I do think it would be more involved and require extra infrastructure (infra costs ++).

This seems a good use case for the pluggable load balancer initiative that was brought up today.

> Right now what I'm thinking of doing is using the iptables hack to unblock this PR and then work on an upstream proposal to achieve 4) as a long term solution.

+1

File: templates/flavors/private/patches/ilb-iptables-cmd.yaml
Comment (CecileRobertMichon) [inline templates/flavors/private/patches/ilb-iptables-cmd.yaml:10]: `if [ -f /tmp/kubeadm.yaml ]` should be replaced by https://github.com/kubernetes-sigs/cluster-api/issues/3828 in the future

**Comment by CecileRobertMichon** [comment]
This is ready pending https://github.com/kubernetes/test-infra/pull/19659 to fix CI e2e (and squash)

**Comment by CecileRobertMichon** [comment]
https://github.com/kubernetes/test-infra/pull/19678 merged

/retest

File: cloud/scope/cluster.go
Comment (devigned) [inline cloud/scope/cluster.go:130]:
```suggestion
	specs := []azure.LBSpec{
```

File: cloud/scope/cluster.go
Comment (devigned) [inline cloud/scope/cluster.go:158]: Perhaps, invert this.

``` golang
if !s.IsAPIServerPrivate() {
    return specs
}

// do private stuff
return specs
```

File: cloud/scope/cluster.go
Comment (devigned) [inline cloud/scope/cluster.go:484]: Perhaps, the following:

```golang
if !s.IsAPIServerPrivate() && s.APIServerPublicIP().DNSName == "" {
    //...
}

**Review by devigned — CHANGES_REQUESTED** [review]
Couple nits... The PR is really big, so I think we'd probably judge it mostly on tests.

Overall it looks like great work.

**Comment by CecileRobertMichon** [comment]
MachinePool test faied but private cluster test passed 😢 

/retest

**Comment by CecileRobertMichon** [comment]
I did some squashing with what is hopefully a logical breakdown of commits. E2E test is now passing with the addition of the iptables workaround. 

@devigned and I did some thinking about how we could set up an [Azure Private DNS Zone](https://docs.microsoft.com/en-us/azure/dns/private-dns-overview) and have a hosts entry for the apiserver DNS to redirect to localhost in each control plane. I think that would be more elegant than needing the iptables rule and would solve the hairpin routing on all control plane VMs (not just for kubeadm init). Unfortunately, private dns zones are not supported in the Go SDK yet (https://github.com/Azure/azure-sdk-for-go/issues/12965) so I think we should try to merge this PR as-is with iptables hack and switch to using private DNS as soon as it's available in the SDK.

@nader-ziada @mboersma @alexeldeib please review when you can, these are extensive changes so I'd like to get as many 👀  on it as possible. Let me know if there is any part that is unclear I'm happy to pair on review and walk through code.

/hold cancel

File: config/crd/bases/infrastructure.cluster.x-k8s.io_azureclusters.yaml
Comment (nader-ziada) [inline config/crd/bases/infrastructure.cluster.x-k8s.io_azureclusters.yaml:530]: seems like an extra `'`

File: docs/book/src/topics/api-server-endpoint.md
Comment (nader-ziada) [inline docs/book/src/topics/api-server-endpoint.md:21]: maybe add a link to how to use with [bastion](https://docs.microsoft.com/en-us/azure/bastion/bastion-overview)?

File: templates/cluster-template-private.yaml
Comment (nader-ziada) [inline templates/cluster-template-private.yaml:33]: couldn't find the docs any mention to `AZURE_INTERNAL_LB_IP ` as an env variable

File: templates/cluster-template-private.yaml
Comment (nader-ziada) [inline templates/cluster-template-private.yaml:3]: When I tried running `make create-workload-cluster` with `CLUSTER_TEMPLATE=cluster-template-private.yaml` I kept getting the following error
```
error: error validating "STDIN": error validating data: [ValidationError(AzureCluster.spec.networkSpec.subnets[0].cidrBlocks): unknown object type "nil" in AzureCluster.spec.networkSpec.subnets[0].cidrBlocks[0], ValidationError(AzureCluster.spec.networkSpec.subnets[1].cidrBlocks): unknown object type "nil" in AzureCluster.spec.networkSpec.subnets[1].cidrBlocks[0], ValidationError(AzureCluster.spec.networkSpec.vnet): missing required field "name" in io.x-k8s.cluster.infrastructure.v1alpha3.AzureCluster.spec.networkSpec.vnet]; if you choose to ignore these errors, turn validation off with --validate=false
make: *** [create-workload-cluster] Error 1
```
even though the regular template worked fine, maybe something with the validation, but didn't find it just yet. Did you try creating the workload cluster this way?

File: api/v1alpha3/types.go
Comment (CecileRobertMichon) [reply api/v1alpha3/types.go:148]: That's a good question. From an Azure perspective, a load balancer can have more than 1 frontend IP. For the API Server LB, we only allow one frontend IP currently but my reasoning for making the type an array was that 1) we might need more than 1 in the future and 2) we might reuse the LoadBalancerSpec type for other LoadBalancer specs in the future so it's easier to make it an array from the start so we don't have to make API changes later. I agree it makes the UX a little awkward because an array implies you can specify more than one. What do you think?

File: config/crd/bases/infrastructure.cluster.x-k8s.io_azureclusters.yaml
Comment (CecileRobertMichon) [reply config/crd/bases/infrastructure.cluster.x-k8s.io_azureclusters.yaml:530]: I think that's how deprecated fields are marked? There is no `'` in the type description, but looks like cidrBlock description is also in single quotes

File: api/v1alpha3/types.go
Comment (nader-ziada) [reply api/v1alpha3/types.go:148]: matching the azure types is a good reason, its fine to keep it an array as long as we have a good error message in the validation and document it somewhere (maybe a comment on the type).

File: templates/cluster-template-private.yaml
Comment (CecileRobertMichon) [reply templates/cluster-template-private.yaml]: removed, was only using that for the test so I made it a patch on the test template instead. The privateIP field is already documented.

File: templates/cluster-template-private.yaml
Comment (CecileRobertMichon) [reply templates/cluster-template-private.yaml:3]: That's because you were missing some env vars for subnet CIDRs. I removed all the env vars I added in the template for the test so validation should work now. Keep in mind that your management cluster needs to be in the same vnet for this to work though so it won't work with `make create-workload-cluster` if you're using kind as your management cluster.

File: docs/book/src/topics/api-server-endpoint.md
Comment (CecileRobertMichon) [reply docs/book/src/topics/api-server-endpoint.md:21]: I'll hold until we have https://github.com/kubernetes-sigs/cluster-api-provider-azure/issues/165 and have tested Bastion, I don't want to guide users to use it until we've verified it works well with capz.

File: api/v1alpha3/azurecluster_default_test.go
Comment (CecileRobertMichon) [reply api/v1alpha3/azurecluster_default_test.go:547]: added

File: cloud/scope/managedcontrolplane.go
Comment (alexeldeib) [inline cloud/scope/managedcontrolplane.go:229]: nit: actually `"kubernetes"`

but shouldn't be provisioned by CAPZ, only used for machines (not relevant to this pr)

**Comment by CecileRobertMichon** [comment]
/retest

ipv6 flake :(

File: api/v1alpha3/azurecluster_validation.go
Comment (alexeldeib) [inline api/v1alpha3/azurecluster_validation.go:210]: if we only support standard we should remove https://github.com/kubernetes-sigs/cluster-api-provider-azure/blob/3c5cb5b975548137d1675bf685766d624bcfa48b/api/v1alpha3/types.go#L174-L175

Do we even need the field in the spec? it's hardcoded in the service right now: https://github.com/kubernetes-sigs/cluster-api-provider-azure/blob/master/cloud/services/loadbalancers/loadbalancers.go#L78

File: cloud/scope/cluster.go
Comment (alexeldeib) [inline cloud/scope/cluster.go:153]: n.b.: https://github.com/kubernetes-sigs/cluster-api-provider-azure/compare/master...alexeldeib:ace/integrate?expand=1#diff-fcb822faaca21eede09f649536848e46d7789e73a2ba9f2ad1b50ebc1d298ed7R167

https://github.com/kubernetes-sigs/cluster-api-provider-azure/compare/master...alexeldeib:ace/integrate?expand=1#diff-bfd4ab4960c565313b05443fc66a48e60a79f9866f6a39fad03044b2161ac69eR192-R194

It'll make my life a little smoother in the future if you don't use generate + LB name on the scope, here but attach the pool name directly to the scope :)

**Review by alexeldeib — COMMENTED** [review]
I took a pass over this, looks sane to me. Couple minor comments

**Comment by alexeldeib** [comment]
this is a big pr, nice work :)

File: api/v1alpha3/azurecluster_default.go
Comment (CecileRobertMichon) [reply api/v1alpha3/azurecluster_default.go:42]: I'm not actually using it so I can remove it, I moved it from somewhere else and didn't know if I was going to need it later

File: api/v1alpha3/azurecluster_default_test.go
Comment (CecileRobertMichon) [reply api/v1alpha3/azurecluster_default_test.go:586]: I was sure I pushed an update to this test 😕 I wonder if I lost it in the rebase

File: api/v1alpha3/azurecluster_validation.go
Comment (CecileRobertMichon) [reply api/v1alpha3/azurecluster_validation.go:210]: we technically don't, I added it as mostly informative and with the possibility that we might support more than one later in mind

good catch though, I should definitely use it if I keep it 😬

File: cloud/scope/cluster.go
Comment (CecileRobertMichon) [reply cloud/scope/cluster.go]: I moved it to the describer but didn't take it out of scope since everything else is there right now, we can lift and shift everything as part of #825

**Comment by CecileRobertMichon** [comment]
@alexeldeib I think I've addressed all your comments, will let you take a look at the added commits before I squash 🎃

File: test/e2e/azure_test.go
Comment (devigned) [inline test/e2e/azure_test.go:78]: Hmm... Did you consider:

``` go
if os.Getenv("LOCAL_ONLY") != "true" {
    Skip("test requires deployment to Azure")
}
```

**Review by devigned — APPROVED** [review]
/lgtm

/assign @alexeldeib

**Comment by nader-ziada** [comment]
🎉  great work Cecile 
lgtm

**Comment by alexeldeib** [comment]
is this the hairpin issue? looks like a timeout hitting the api server with 3 master setup? didn't look too deeply I admit 


/test pull-cluster-api-provider-azure-e2e

```
 Unexpected error:
      <*url.Error | 0xc0004d8090>: {
          Op: "Delete",
          URL: "https://capz-e2e-b3o4h8-a7437409.westus2.cloudapp.azure.com:6443/api/v1/namespaces/default/services/httpd-ilb",
          Err: {
              Op: "read",
              Net: "tcp",
              Source: {IP: [10, 60, 5, 156], Port: 44610, Zone: ""},
              Addr: {IP: [20, 69, 81, 80], Port: 6443, Zone: ""},
              Err: {Syscall: "read", Err: 0x6e},
          },
      }
      Delete https://capz-e2e-b3o4h8-a7437409.westus2.cloudapp.azure.com:6443/api/v1/namespaces/default/services/httpd-ilb: read tcp 10.60.5.156:44610->20.69.81.80:6443: read: connection timed out
  occurred
```

**Comment by CecileRobertMichon** [comment]
> is this the hairpin issue? looks like a timeout hitting the api server with 3 master setup? didn't look too deeply I admit

no it was the HA public cluster spec that failed `capz-e2e: Workload cluster creation With 3 control-plane nodes and 2 worker nodes`, looks like a flake

**Comment by CecileRobertMichon** [comment]
@alexeldeib let me know when you're okay with me rebasing

**Comment by CecileRobertMichon** [comment]
Rebased + squashed

@alexeldeib @devigned @nader-ziada ready for a final look

**Review by devigned — APPROVED** [review]
/lgtm

**Pull request merged**
